### ISSUE TITLE
Question: Resource (User/Group) ID substitution in scim gateway

### ISSUE THREAD

**Issue by j--wong** [jelhub/scimgateway#65]
Hi,

We are looking at using scimgateway to expose SCIM 2.0 interface, and are building a custom plugin to integrate with our own identity service backend.  We have a question around a piece of code we came across as we develop our plugin. 

When we create a user record in our plugin, our plugin returns the full user object (including the ID) of the newly created user in the response. However, this piece of code substitute our `id` with the `userName` - `jsonBody.id = jsonBody.userName`. And do another `GET` on the same object using the `userName` field.

https://github.com/jelhub/scimgateway/blob/master/lib/scimgateway.js#L1094-L1101

If `userName` doesn't exist on the return object, it would do the same with `externalId` field.

Any reasons the scimgateway needs to:
1. substitute the `id` returned by the plugin with either `userName` or `externalId` and
2. re-retrieve the same object using a different filter

While the plugin returns full user object.

We are just wondering if there's a usecase that scimgateway is catering for that we are not aware of. 

Thanks,
Josh

**Comment by jelhub** [comment]
Hi,
getUser() should support returning id, and we have to use this method for adding the id to the full user object. If id not supported, it will be set to userName/externalID.

A1)
Returned id will be use and not substituted.
Ref: https://github.com/jelhub/scimgateway/blob/master/lib/scimgateway.js#L1112
Seems your getUser() does not return the id

A2)
Not sure I understand this question?

Regards,
Jarle

**Comment by jelhub** [comment]
I see that code could have been improved by skipping getUser() if the the plugin createUser() do return an object that includes the id.

ref: https://github.com/jelhub/scimgateway/blob/master/lib/scimgateway.js#L1087-L1090

That's probably what you meant with Q2?

Jarle

**Comment by j--wong** [comment]
Thanks for the reply.  Yes, if we skip the getUser() call if the object (returned by the plugin) already contains the `id` would fix the issue we are encountering.

Would you like me to PR that in? I would just put in a check like this:

```javascript
if (!jsonBody.id) {
  let obj
  try { // retrieve id
    if (handle.createMethod === 'createUser') {
      if (jsonBody.userName) {
        jsonBody.id = jsonBody.userName
        obj = await this[handle.getMethod](ctx.params.baseEntity, { attribute: 'userName', operator: 'eq', value: jsonBody.userName }, ['id', 'userName'])
      } else if (jsonBody.externalId) {
        jsonBody.id = jsonBody.externalId
        obj = await this[handle.getMethod](ctx.params.baseEntity, { attribute: 'externalId', operator: 'eq', value: jsonBody.externalId }, ['id', 'externalId'])
      }
    } else if (handle.createMethod === 'createGroup') {
      if (jsonBody.externalId) {
        jsonBody.id = jsonBody.externalId
        obj = await this[handle.getMethod](ctx.params.baseEntity, { attribute: 'externalId', operator: 'eq', value: jsonBody.externalId }, ['id', 'externalId'])
      } else if (jsonBody.displayName) {
        jsonBody.id = jsonBody.displayName
        obj = await this[handle.getMethod](ctx.params.baseEntity, { attribute: 'displayName', operator: 'eq', value: jsonBody.displayName }, ['id', 'displayName'])
      }
    }
  } catch (err) { }
  if (obj && obj.id) jsonBody.id = obj.id
}
```

**Comment by jelhub** [comment]
PR not needed.
Will be included in next update

Thanks,
Jarle

**Comment by jelhub** [comment]
It's now fixed in v4.1.3

**Comment by j--wong** [comment]
Amazing, thank you @jelhub
